_posts/2023-06-19-End-to-end-acknowledgements-in-Data-Prepper.md

```diff
@@ -0,0 +1,64 @@
+
+<img src="/assets/media/blog-images/2023-06-16-end-to-end-acknowledgements/end-to-end-acknowledgements.png" alt="End-to-end acknowledgments diagram"/>{: .img-fluid}
+
+
+1. Data Prepper source receives a batch of records from an external source (such as Amazon S3).
+2. When a Data Prepper source is configured with end-to-end acknowledgments, the source creates an AcknowledgementSet along with a callback function for each batch of records received from the external source. The Source converts the records to Data Prepper events
+3. Each event in the batch is added to the AcknowledgementSet, and a reference to the AcknowledgementSet is kept in the event as a handle
+4. Events are then passed to the ingestion pipeline, which consists of multiple processors that transform, filter, enrich the data in the event
+5. If an event is dropped in the processors, the AcknowledgementSet is notified of the completion of the event when the event handle is released
+6. All the events that are not dropped are sent to the Data Prepper sink
+7. The Data Prepper sink sends the events to the external sink (such as OpenSearch Cluster)
+8. When events are successfully sent to the external sink, the AcknowledgementSet is notified of the event's completion when the event handles are released.
+9. When the AcknowledgementSet has no pending events, it invokes the callback function
+10. The callback function notifies the external source about the successful delivery of the batch of records
+
+If all events are not successfully delivered to the external sinks, AcknowledgementSet will retain some events that have not been released. Each AcknowledgementSet has an expiry time, and if all events are not released before the expiry time, the AcknowledgementSet is freed, and external source is not notified of the delivery of the records. This situation may trigger Data Prepper source or the external source to take corrective action, such as retrying the ingestion of the records again.
+
+Data Prepper supports multiple pipelines and events can be routed to different pipelines based on the configuration and the data within the events. Additionally, each event may be routed to multiple pipelines that eventually send data to different sinks. However, in some cases, multiple copies of the same event are not created when it is routed to multiple pipelines/sinks. This poses a challenge when tracking acknowledgments because multiple acknowledgments are required for the same event. To address this issue, reference counts are maintained for events that are routed to multiple pipelines. When an event handle is released (either in the step 5 or step 8), the reference count is decremented. The event is considered completed and the handle is removed from the AcknowledgementSet only when the reference count reaches zero, indicating the successful delivery of the event to all sinks. This approach ensures accurate tracking of acknowledgments for events routed to multiple pipelines.
```

Comment:
```suggestion
Data Prepper supports multiple pipelines, and events can be routed to different pipelines based on the configuration and the data within the events. Additionally, each event may be routed to multiple pipelines that eventually send data to different sinks. In some cases, multiple copies of the same event are not created when the event is routed to multiple pipelines or sinks. This poses a challenge when tracking acknowledgments because multiple acknowledgments are required for the same event. To address this issue, reference counts are maintained for events that are routed to multiple pipelines. When an event handle is released (either in step 5 or step 8), the reference count is decremented. The event is considered completed, and the handle is removed from the AcknowledgmentSet only when the reference count reaches zero, indicating the successful delivery of the event to all sinks. This approach ensures accurate tracking of acknowledgments for events routed to multiple pipelines.
```